enterprise/backend/src/metabase_enterprise/advanced_permissions/models/permissions/group_manager.clj

```diff
@@ -0,0 +1,56 @@
+(ns metabase-enterprise.advanced-permissions.models.permissions.group-manager
+  (:require [clojure.data :as data]
+            [clojure.set :refer [subset?]]
+            [metabase.api.common :as api]
+            [metabase.models :refer [PermissionsGroupMembership]]
+            [metabase.util :as u]
+            [metabase.util.i18n :refer [tru]]
+            [toucan.db :as db]))
+
+(defn user-group-memberships
+  "Return a list of group memberships a User belongs to.
+  Group Membership is a map with 2 keys [:id :is_group_manager]."
+  [user-or-id]
+  (when user-or-id
+    (db/select [PermissionsGroupMembership [:group_id :id] :is_group_manager] :user_id (u/the-id user-or-id))))
+
+(defn- user-group-memberships->map
+  "Transorm user-group-memberships to a map with keys are group-id and values are membership info.
+
+  [{:id 1, :is_group_manager true}] => {1 {:is_group_manager true}}
+
+  We can diff this map to decide which membership to adds/removes."
+  [user-group-memberships]
+  (into {} (map (fn [x] [(:id x) (dissoc x :id)]) user-group-memberships)))
+
+(defn set-user-group-memberships!
+  "Update User Groups Memberships when `advanced-permissions` is enabled.
+  It's used to add/remove users from groups and also promote/demote Group Managers."
+  [user-or-id new-user-group-memberships]
+  (let [user-id                       (u/the-id user-or-id)
+        old-user-group-memberships    (user-group-memberships user-id)
+        old-group-id->membership-info (user-group-memberships->map old-user-group-memberships)
+        new-group-id->membership-info (user-group-memberships->map new-user-group-memberships)
+        [to-remove to-add]            (data/diff old-group-id->membership-info new-group-id->membership-info)
+        to-remove-group-ids           (keys to-remove)
+        to-add-group-ids              (keys to-add)]
+    ;; TODO: Should do this check in the API layer
+    (when-not api/*is-superuser?*
+      ;; prevent groups manager from update membership of groups that they're not manager of
+      (when-not (and api/*is-group-manager?*
+                     (subset? (set (concat to-remove-group-ids to-add-group-ids))
+                              (db/select-field :group_id PermissionsGroupMembership
+                                               :user_id api/*current-user-id* :is_group_manager true)))
+        (throw (ex-info (tru "Not allowed to edit group memberships")
+                        {:status-code 403}))))
+    (when (seq (concat to-remove-group-ids to-add-group-ids))
```

Comment:
nit: haven't timed it, but if both group ID lists are large, might be more efficient to skip the `concat` and just call `seq` on each one separately?
```suggestion
    (when (or (seq to-remove-group-ids) (seq to-add-group-ids))
```

Reply:
better yet, do you need this conditional at all? both DB updates below are already gated by conditionals